### PR TITLE
rpc: Move the `generate` RPC call to rpcwallet

### PR DESCRIPTION
**Pull request by laanwj**
This makes it possible to mine to any wallet when multi-wallet mode is added. Solves the same problem as #10649, but IMO in a cleaner way.

It also gets rid of the circuitous `ScriptForMining` method on `CValidationInterface`, which really doesn't belong there.

After this change it's still possible to mine without wallet through `generatetoaddress`. I first proposed this in #7965.

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
...
+            "1. nblocks      (numeric, required) How many blocks are generated immediately.\n"
+            "2. maxtries     (numeric, optional) How many iterations to try (default = 1000000).\n"
+            "\nResult:\n"
+            "[ blockhashes ]     (array) hashes of blocks generated\n"
+            "\nExamples:\n"
+            "\nGenerate 11 blocks\n"
+            + HelpExampleCli("generate", "11")
+        );
+
+    int nGenerate = request.params[0].get_int();
```
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: nit: variable name should be lowercase, not hungarian

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
...
+    uint64_t nMaxTries = 1000000;
```
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: nit: variable name should be lowercase, not hungarian

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
...
+    if (request.params.size() > 1) {
```
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: nit: should handle case when `request.params[1]` is Null.

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
...
+        nMaxTries = request.params[1].get_int();
+    }
+
+    std::shared_ptr<CReserveScript> coinbaseScript;
+    pwallet->GetScriptForMining(coinbaseScript);
+
+    // If the keypool is exhausted, no script is returned at all.  Catch this.
+    if (!coinbaseScript)
```
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: nit: braces

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
...
+        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
+
+    //throw an error if no script was provided
+    if (coinbaseScript->reserveScript.empty())
```
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: nit: braces

**Review by jnewbery — COMMENTED** [review]
tested ACK 8c410a8fbb5f1876484a3f10d25cf9c06a8ef875. This is a sensible change. Anything that moves us towards #7965 and away from circular libbitcoin_server.a/libbitcoin_wallet.a dependencies is a good thing in my book.

I realise that this is mostly code-move. There are a bunch of style nits in the moved code - entirely up to you if you want to address them as part of this PR or leave them as they are.

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/10683/commits/8c410a8fbb5f1876484a3f10d25cf9c06a8ef875 though nit cleanup is great too.

**Comment by jonasschnelli** [comment]
Nice and thanks for doing this:
utACK 8c410a8fbb5f1876484a3f10d25cf9c06a8ef875

**Comment by laanwj** [comment]
Ok ok I'll add a commit to clean up the style nits, don't think it should be done at the time of the move as that just complicates review.

**Comment by jnewbery** [comment]
utACK 4cdb6209de2f712cd5847373b6bc2fb6401fb739. No more nits. Thanks!

> don't think it should be done at the time of the move as that just complicates review.

I'd agree if the nit fixes were in the same commit, but if they're done in their own commit either before or after the move it adds almost no review burden (as long as reviewers are reviewing commitwise)

**Comment by laanwj** [comment]
> I'd agree if the nit fixes were in the same commit, but if they're done in their own commit either before or after the move it adds almost no review burden (as long as reviewers are reviewing commitwise)

Yes that's what I meant, with a separate commit it's fine.

File: src/rpc/mining.h [hunk 6]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_RPC_MINING_H
+#define BITCOIN_RPC_MINING_H
+
+#include <script/script.h>
```
Comment (TheBlueMatt) [inline src/rpc/mining.h]: Shouldnt this be a "-style include?

**Review by TheBlueMatt — COMMENTED** [review]
utACK 4cdb6209de2f712cd5847373b6bc2fb6401fb739, thanks for cleaning up CValidationInterface.

File: src/rpc/mining.h (hunk 6 above)
Comment (laanwj) [inline src/rpc/mining.h]: Yes, good point

**Comment by laanwj** [comment]
Updated `#include` style and squashed this into the first commit, where `mining.h` was introduced:
8c410a8fbb5f1876484a3f10d25cf9c06a8ef875 → df7e2f057b6c9f0f7c950f9077dc63a577f54117
4cdb6209de2f712cd5847373b6bc2fb6401fb739 → 2a962834febc9f56126ef06cf1bd5e1b02370278

**Pull request merged**